### PR TITLE
Validations/joi

### PR DESCRIPTION
**Pull request by jaimin45**

### CODE REVIEW

**Review by JainishaParmar — APPROVED** [review]

**Review by JainishaParmar — APPROVED** [review]

File: routes/users.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+const express = require("express");
```
Comment (kapilbarad) [inline routes/users.js]: What is this file for?

File: routes/index.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+const express = require("express");
```
Comment (kapilbarad) [inline routes/index.js]: Why we've added this fine again?

File: validations/tutorialsValidate.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+const joi = require("joi");
+
+const schema = joi.object({
+  title: joi.string().min(3).max(100).required(),
+  description: joi.string().min(1).max(5000).required(),
+  published: joi.boolean().required(),
+});
+
+module.exports = {
+  schema,
```
Comment (kapilbarad) [inline validations/tutorialsValidate.js]: You should name it in a way so it won't conflict with other variables, Like `postTutorialSchema` and `patchTutorialSchema`.

File: controllers/tutorialsControllers.js [hunk 4]
```diff
@@ -36,6 +38,7 @@ const updateTutorial = async (req, res) => {
     if (typeof title !== "undefined") update.title = title;
     if (typeof description !== "undefined") update.description = description;
     if (typeof published !== "undefined") update.published = published;
+    await schema.validateAsync(req.body);
```
Comment (kapilbarad) [inline controllers/tutorialsControllers.js]: Did you test by passing invalid title or any other fields? What's the response you're getting?

File: validations/tutorialsValidate.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+const joi = require("joi");
```
Comment (kapilbarad) [inline validations/tutorialsValidate.js]: You can use `tutorial.validations.js` or just `tutorial.js` as the file name.

File: controllers/tutorialsControllers.js [hunk 6]
```diff
@@ -21,6 +22,7 @@ const createTutorial = async (req, res) => {
     published: req.body.published,
   });
   try {
+    await schema.validateAsync(req.body);
```
Comment (kapilbarad) [inline controllers/tutorialsControllers.js]: Did you test by passing invalid title or any other fields? What's the response you're getting?

File: controllers/tutorialsControllers.js (hunk 4 above)
Comment (kapilbarad) [reply controllers/tutorialsControllers.js]: This is how you should be using Joi to validate:
```
const { error, body: tutorialBody } = postTutorialSchema.validate(req.body);

# Return error response if there is anything in error variable
...
...
...
const tutorial = await Tutorial.findByIdAndUpdate(
      { _id: req.params.id },
      tutorialBody,
      { new: true }
    );
```

**Review by kapilbarad — CHANGES_REQUESTED** [review]

**Review by JainishaParmar — APPROVED** [review]

File: controllers/tutorialsControllers.js (hunk 4 above)
Comment (kapilbarad) [reply controllers/tutorialsControllers.js]: @jaimin45 , Please follow this example that I shared above.

File: controllers/tutorialsControllers.js [hunk 7]
```diff
@@ -21,27 +22,33 @@ const createTutorial = async (req, res) => {
...
+      res.status(500).send({ message: "Internal Server Error" });
+    }
   }
 };
 
 // Update a Tutorial by ID
 const updateTutorial = async (req, res) => {
   try {
     const { title, description, published } = req.body;
+    const { error, body: tutorialBody } = postTutorialSchema.validate(req.body);
```
Comment (kapilbarad) [inline controllers/tutorialsControllers.js]: We should return the error here, no need to always return error from catch block. 
```
if (error) {
   return res.status(400).send({ message: <error-detail> });
}
```

Above code will return 400 error if there will be any validation errors.

CC: @JainishaParmar @ashishkava

File: controllers/tutorialsControllers.js [hunk 8]
```diff
@@ -21,27 +22,33 @@ const createTutorial = async (req, res) => {
...
     if (typeof description !== "undefined") update.description = description;
     if (typeof published !== "undefined") update.published = published;
     const tutorial = await Tutorial.findByIdAndUpdate(
       { _id: req.params.id },
       update,
-      {
-        new: true,
-      }
+      tutorialBody,
+      error,
```
Comment (kapilbarad) [inline controllers/tutorialsControllers.js]: Why are we passing error here?

**Review by kapilbarad — CHANGES_REQUESTED** [review]

**Review by JainishaParmar — APPROVED** [review]

**Review by ghost — APPROVED** [review]

File: controllers/tutorialsControllers.js [hunk 9]
```diff
@@ -21,41 +22,47 @@ const createTutorial = async (req, res) => {
...
+    const { error } = postTutorialSchema.validate(req.body);
+    if (error) {
+      return res.status(400).send({ message: error.message });
+    }
+    await postTutorialSchema.validateAsync(req.body);
     await tutorial.save();
     res.status(201).send();
   } catch (error) {
-    res.status(500).send({ message: "Internal Server Error" });
+    if (error) {
```
Comment (kapilbarad) [inline controllers/tutorialsControllers.js:33]: Please verify CastError same as postTutorial API

**Review by kapilbarad — CHANGES_REQUESTED** [review]

**Review by JainishaParmar — APPROVED** [review]

**Comment by jaimin45** [comment]
Validations/Joi

**Pull request merged**